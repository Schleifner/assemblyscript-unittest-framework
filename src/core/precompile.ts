/**
 * Will transform all source file to get all relative functions
 */

import ignore from "ignore";
import { fileURLToPath, URL } from "node:url";
import { main } from "assemblyscript/asc";
import { join, relative, resolve } from "node:path";
import { getIncludeFiles } from "../utils/pathResolver.js";
import { SourceFunctionInfo, UnittestPackage } from "../interface.js";

const sourceFunctions = new Map<string, SourceFunctionInfo[]>();
const listFunctions = join(fileURLToPath(new URL(".", import.meta.url)), "..", "transform", "listFunctions.mjs");

export async function precompile(
  includes: string[],
  excludes: string[],
  testcases: string[] | undefined,
  transformFunction = listFunctions
): Promise<UnittestPackage> {
  // if specify testcases, use testcases for unittest
  // otherwise, get testcases(*.test.ts) in includes directory
  const testCodePaths = testcases ?? getRelatedFiles(includes, excludes, (path: string) => path.endsWith(".test.ts"));

  const sourceCodePaths = getRelatedFiles(includes, excludes, (path: string) => !path.endsWith(".test.ts"));
  const sourceCodeTransfroms: Promise<void>[] = [];
  for (const sourceCodePath of sourceCodePaths.values()) {
    sourceCodeTransfroms.push(transform(sourceCodePath, transformFunction));
  }
  await Promise.all(sourceCodeTransfroms);

  return {
    testCodePaths,
    sourceFunctions,
  };
}

// a. include in config
// b. exclude in config
export function getRelatedFiles(includes: string[], excludes: string[], filter: (path: string) => boolean) {
  const result: string[] = [];
  const includeFiles = getIncludeFiles(includes, (path) => path.endsWith(".ts")); // a
  const exc = ignore.default().add(excludes);

  for (const path of includeFiles) {
    const relativedPath = relative(".", path);
    if (relativedPath.startsWith("..")) {
      throw new Error(`file ${path} out of scope (${resolve(".")})`);
    }
    if (exc.ignores(relativedPath)) {
      continue; // ab
    }
    if (filter(path)) {
      result.push(path.replaceAll(/\\/g, "/"));
    }
  }
  return result;
}

async function transform(sourceCodePath: string, transformFunction: string) {
  const { error, stderr } = await main([
    sourceCodePath,
    "--noEmit",
    "--disableWarning",
    "--transform",
    transformFunction,
  ]);
  if (error) {
    // eslint-disable-next-line @typescript-eslint/no-base-to-string
    console.error(stderr.toString());
    throw error;
  }
  sourceFunctions.set(sourceCodePath, functionInfos as SourceFunctionInfo[]);
}
