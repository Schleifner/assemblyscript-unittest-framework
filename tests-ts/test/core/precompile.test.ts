import { join } from "node:path";
import { fileURLToPath, URL } from "node:url";
import { precompile } from "../../../src/core/precompile.js";

test("listFunction transform", async () => {
  const transformFunction = join(
    fileURLToPath(new URL(".", import.meta.url)),
    "..",
    "..",
    "..",
    "transform",
    "listFunctions.mjs"
  );
  const unittestPackages = await precompile(["tests-ts/fixture/transformFunction.ts"], [], [], transformFunction);
  expect(unittestPackages.testCodePaths).toEqual([]);
  expect(unittestPackages.sourceFunctions).toMatchSnapshot();
});
