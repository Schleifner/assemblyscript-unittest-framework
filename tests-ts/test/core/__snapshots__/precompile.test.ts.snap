// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listFunction transform 1`] = `
Map {
  "tests-ts/fixture/transformFunction.ts" => [
    {
      "name": "tests-ts/fixture/transformFunction/Foo.bar",
      "range": [
        49,
        49,
      ],
    },
    {
      "name": "tests-ts/fixture/transformFunction/Foo#constructor",
      "range": [
        45,
        45,
      ],
    },
    {
      "name": "tests-ts/fixture/transformFunction/Foo#foo",
      "range": [
        42,
        42,
      ],
    },
    {
      "name": "",
      "range": [
        33,
        33,
      ],
    },
    {
      "name": "tests-ts/fixture/transformFunction/subscribe",
      "range": [
        28,
        28,
      ],
    },
    {
      "name": "b",
      "range": [
        20,
        22,
      ],
    },
    {
      "name": "tests-ts/fixture/transformFunction/a4",
      "range": [
        20,
        22,
      ],
    },
    {
      "name": "",
      "range": [
        16,
        16,
      ],
    },
    {
      "name": "",
      "range": [
        14,
        14,
      ],
    },
    {
      "name": "",
      "range": [
        12,
        12,
      ],
    },
    {
      "name": "",
      "range": [
        12,
        12,
      ],
    },
    {
      "name": "",
      "range": [
        4,
        4,
      ],
    },
    {
      "name": "",
      "range": [
        3,
        8,
      ],
    },
  ],
}
`;
